timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,336 @@
+peek ::
+  Timeline a ->
+  -- | The time to peek. Any valid 'UTCTime' value can be passed in.
+  UTCTime ->
+  a
+peek Timeline {..} time = maybe initialValue snd $ Map.lookupLE time values
+
+-- | A time range. Each bound is optional. 'Nothing' represents infinity.
+data TimeRange = TimeRange
+  { -- | inclusive
+    from :: Maybe UTCTime,
+    -- | exclusive
+    to :: Maybe UTCTime
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+-- | If all time in 'TimeRange' is less than the given 'UTCTime'
+isTimeAfterRange :: UTCTime -> TimeRange -> Bool
+isTimeAfterRange t TimeRange {to} = maybe False (t >=) to
+
+instance FunctorWithIndex TimeRange Timeline where
+  imap :: (TimeRange -> a -> b) -> Timeline a -> Timeline b
+  imap f Timeline {..} =
+    Timeline
+      { initialValue = f initialRange initialValue,
+        values = flip Map.mapWithKey values $ \from value ->
+          let timeRange = TimeRange (Just from) (fst <$> Map.lookupGT from values)
+           in f timeRange value
+      }
+    where
+      initialRange = TimeRange Nothing $ fst <$> Map.lookupMin values
+
+instance FoldableWithIndex TimeRange Timeline
+
+instance TraversableWithIndex TimeRange Timeline where
+  itraverse :: (Applicative f) => (TimeRange -> a -> f b) -> Timeline a -> f (Timeline b)
+  itraverse f = sequenceA . imap f
+
+-- | Return a set of 'UTCTime's when the value changes
+changes :: Timeline a -> Set UTCTime
+changes Timeline {values} = Map.keysSet values
+
+-- | A value with @effectiveFrom@ and @effectiveTo@ attached. This is often the
+-- type we get from inputs. A list of @'Record' a@ can be converted to
+-- @'Timeline' ('Maybe' a)@. See 'fromRecords'.
+data Record a = Record
+  { -- | inclusive
+    from :: UTCTime,
+    -- | exclusive. When 'Nothing', the record never expires, until there is
+    -- another record with a newer 'effectiveFrom' time.
+    to :: Maybe UTCTime,
+    value :: a
+  }
+  deriving stock (Show, Eq, Functor, Foldable, Traversable)
+
+-- | Get the "effective from" time
+recordFrom :: Record a -> UTCTime
+recordFrom Record {from} = from
+
+-- | Get the "effective to" time
+recordTo :: Record a -> Maybe UTCTime
+recordTo Record {to} = to
+
+-- | Get the value wrapped in a @'Record' a@
+recordValue :: Record a -> a
+recordValue = value
+
+-- | A smart constructor for @'Record' a@.
+-- Returns 'Nothing' if @effectiveTo@ is not greater than @effectiveFrom@
+makeRecord ::
+  -- | effective from
+  UTCTime ->
+  -- | optional effective to
+  Maybe UTCTime ->
+  -- | value
+  a ->
+  Maybe (Record a)
+makeRecord from to value =
+  if maybe False (from >=) to
+    then Nothing
+    else Just Record {..}
+
+instance (TH.Lift a) => TH.Lift (Record a) where
+  liftTyped Record {..} =
+    [||
+    Record
+      (unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime from))
+      (fmap unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime <$> to))
+      $$(TH.liftTyped value)
+    ||]
+
+newtype LiftUTCTime = LiftUTCTime UTCTime
```

Comment:
It would be good to document why this exists and the conditions under which we can remove it.